### ISSUE TITLE
Functional proposal for custom components rendering

### ISSUE THREAD

**Issue by sebastianlp** [Cambalab/vue-admin#34]
This may be a `.vue` file. Instead of using template syntax, the file exports only the `<script></script>` part.

So, with this you are in charge of rendering something, with the use of the (createElement) function provided by vue to all component render function.

[Read the doc](https://vuejs.org/v2/guide/render-function.html#createElement-Arguments) to become familiar with the API.

Example:
- BirthDay, Mail, Name and PhoneNumber are simple vue components exported with `<template></template>, <script></script> and <style></style>` tags.
- The render function of `smartRendererComponent` must return a valid createElement object.
- You can use the known properties too: `computed, props, etc.`

`smartRendererComponent.vue`
```js
<script>
import { postForm } from '@/api'
import BirthDay from '@/components/Inputs/BirthDay'
import Mail from '@/components/Inputs/Mail'
import Name from '@/components/Inputs/Name'
import PhoneNumber from '@/components/Inputs/PhoneNumber'

import Vue from 'vue'
import VeeValidate, { Validator } from 'vee-validate'
import es from 'vee-validate/dist/locale/es'

// Add locale helper.
Validator.addLocale(es)
Vue.use(VeeValidate, {
  locale: 'es',
  inject: false
})

export default {
  $validates: true,
  render: function (createElement) {
    let self = this
    let csrfTokenInput = createElement(
      'input',
      {
        attrs: {
          type: 'hidden',
          value: self.csrfToken
        }
      }
    )

    let childInputs = this.options.map(function (element) {
      let toCreateElement = ''

      switch (element.formField.name.toUpperCase()) {
        case 'NOMBRE':
          toCreateElement = Name
          break
        case 'CUMPLEANOS':
          toCreateElement = BirthDay
          break
        case 'MAIL':
          toCreateElement = Mail
          break
        case 'WHATSAPP':
          toCreateElement = PhoneNumber
          break
      }

      return createElement(toCreateElement, {
        props: {
          mysqlType: element.formField.mysql_type,
          inputName: element.formField.name.toUpperCase(),
          placeholder: element.formField.html_placeholder,
          validationRules: element.placeFormFieldValidations,
          formFieldId: element.formField.id
        }
      })
    })

    let vForm = createElement('form', {
      on: {
        submit: function (event) {
          event.preventDefault()

          self.$validator.validateAll().then(result => {
            let data = {
              _csrf: self.csrfToken,
              connectorId: self.$store.state.customer.id,
              responses: []
            }
            data.responses = self.$children.map(function (currentValue) {
              return {
                mysqlType: currentValue.mysqlType,
                formFieldId: currentValue.formFieldId,
                data: currentValue.formData
              }
            })
            postForm(self.$store.state.place.slug, data)
          })
        }
      }
    }, [
      csrfTokenInput,
      childInputs,
      createElement('button', {
        attrs: {
          class: 'button disabled',
          disabled: !self.isSubmitDisabled,
          type: 'submit',
          variant: 'success'
        },
        domProps: {
          innerHTML: 'INGRESAR'
        }
      })
    ]
  )

    return createElement('div', {
      'class': {
        'text-center': true
      }
    }, [vForm])
  },
  computed: {
    isSubmitDisabled () {
      let fields = Object.values(this.fields)

      return fields.every(function (element) {
        return element.valid
      })
    }
  },
  props: {
    csrfToken: {
      type: String,
      required: true
    },
    options: {
      type: Array,
      required: true
    }
  }
}
</script>
```

**Comment by sgobotta** [comment]
@sebastianlp thanks for sharing your idea :) I edited the title a little bit so that we understand what we will be dealing with in this issue. I'll be taking a look as soon as I can!

**Comment by sebastianlp** [comment]
If I can help with something, share thoughts, discuss, whatever, contact me!

**Comment by sgobotta** [comment]
I used this issue and the information provided to start an experimental implementation using functional rendering.
I first created a custom component as a user and tried to wrap key form elements, such as inputs and buttons, with some sort of core *wannabe* components, hoping I could then inject functionality into the native ones chosen by the user (v-text-field, v-btn, from vuetify, etc...). Diving for their children to inject functions on click or input/change events didn't seem to be as easy as I thought, though it doesn't seem to be impossible to implement, but a little dirty. I left a few Vue functional components to keep experimenting, that can work as wrappers for other components provided by a user.
I figured we can pass vue components as props to our Resource core component. All this work led me to build a simple and dynamic api that is kind of injected into components provided by a user for a specific view, Create for example. This approach continues in #42
